### ISSUE TITLE
Postprocess Yolov10 from onnx format

### ISSUE THREAD

**Issue by naufalahnaf17** [ultralytics/ultralytics#16366]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I wondered what was wrong with what I had done
I am trying to preprocess on yolov10 which I have fine-tuned with the facemask dataset which I have downloaded from kaggle

preprocess function

    outputs = np.squeeze(outputs[0])
    rows = outputs.shape[0]

    boxes = []
    scores = []
    class_ids = []

    for i in range(rows):
        class_scores = outputs[i][4]
        if class_scores >= 0.5:
            x,y,w,h = outputs[i][0],outputs[i][1],outputs[i][2],outputs[i][3]
            
            left = int(x - w / 2)
            top = int(y - h / 2)
            width = int(w)
            height = int(h)

            print([left,top,width,height])

            boxes.append([left,top,width,height])

    return boxes

but for some reason the bounding box results produced from the postprocess above are different from 

`model.predict(image_path,save=True)`

**With Model Predict**
![image](https://github.com/user-attachments/assets/dde823ea-be9c-4636-8111-0181347bb084)

**Postprocess Function**
![image](https://github.com/user-attachments/assets/576c5d3a-9785-40ff-a5f1-6ab624a96f81)

Is there anything I missed?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @naufalahnaf17, thank you for reaching out to Ultralytics 🚀! We're excited to help. Please remember this is an automated response, and an Ultralytics engineer will assist you soon.

We suggest exploring our [Docs](https://docs.ultralytics.com) for detailed usage examples including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/). Often, common questions and solutions can be found there.

If this is a 🐛 Bug Report, please share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). This will help us debug more effectively.

For custom training ❓ questions, provide details such as dataset examples and training logs, ensuring alignment with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for more interaction. For real-time chat, visit [Discord](https://ultralytics.com/discord) 🎧. Prefer detailed discussions? Head over to [Discourse](https://community.ultralytics.com). Or participate in our [Subreddit](https://reddit.com/r/ultralytics) community.

## Upgrade

Ensure you're using the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLOv8 is supported in various environments with preinstalled dependencies:

- **Notebooks** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a> <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>
- **Google Cloud**: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

A green badge indicates all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests pass, ensuring correct operation across macOS, Windows, and Ubuntu.

We hope this helps, and look forward to resolving your issue! 😊

**Comment by Y-T-G** [comment]
The output is in `xyxy` format I think.

**Comment by naufalahnaf17** [comment]
hi, thank you for your response

as you said
The output format should be like this:

`x1,y1,x2,y2,conf,labels`

for comparison 

model predict output : 
- cls: tensor([0.])
- conf: tensor([0.9343])
- data: tensor([[102.2374, 212.7556, 263.6473, 359.8716,   0.9343,   0.0000]])
- id: None
- is_track: False
- orig_shape: (448, 448)
- shape: torch.Size([1, 6])
- xywh: tensor([[182.9424, 286.3136, 161.4099, 147.1160]])
- xywhn: tensor([[0.4084, 0.6391, 0.3603, 0.3284]])
- xyxy: tensor([[102.2374, 212.7556, 263.6473, 359.8716]])
- xyxyn: tensor([[0.2282, 0.4749, 0.5885, 0.8033]])

postprocess code and output: 

`
for i in range(rows):
    class_scores = outputs[i][4]
    if class_scores >= 0.9:
        print(outputs[i])
    else : 
        print("No Boxes Detected")
`

output : No Boxes Detected

isn't this strange? There should be one box that appears successfully because the confidence is above 0.9

**Comment by Y-T-G** [comment]
It could be a problem with your preprocessing.

You can follow the example here for preprocessing.

https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-OpenCV-ONNX-Python/main.py

**Comment by naufalahnaf17** [comment]
ok, thanks for answering, I'll try it first and make sure everything is according to the instructions

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions, feel free to ask.

**Comment by curtis18** [comment]
@glenn-jocher  I also have a question about the postprocess of onnx format. I am creating a class that can support both onnx and pt output. But the segment ouput from onnx cannot retrieve the similar output from model.predict. Would you please help?

```python
import onnx
import onnxruntime as ort
from ultralytics import YOLO
import numpy as np
import cv2
import ast
from PIL import Image

class ModelLoader:
    def __init__(self, model_path):
        self.model_path = model_path
        self.use_onnx = model_path.endswith('.onnx')
        self.model = None
        self.class_names = {}
        self.input_size = None
        self.original_image_shape = None
        self.scale_ratio = None
        self.pad = None
        self.load_model()

    def load_model(self):
        if self.use_onnx:
            self.model = ort.InferenceSession(
                self.model_path,
                providers=["CUDAExecutionProvider", "CPUExecutionProvider"]
                if ort.get_device() == "GPU" else ["CPUExecutionProvider"]
            )
            # Extract class names and input size from model metadata
            onnx_model = onnx.load(self.model_path)
            print("Input Name:", self.model.get_inputs()[0].name)
            print("Model Meta:", onnx_model.metadata_props)
            for prop in onnx_model.metadata_props:
                if prop.key == 'names':
                    self.class_names = ast.literal_eval(prop.value)
                if prop.key == 'imgsz':
                    self.input_size = ast.literal_eval(prop.value)[0]
        else:
            # Handle YOLO model loading
            self.model = YOLO(self.model_path)
            self.class_names = self.model.names
            # save_pt_vars_to_txt(model)
            if hasattr(self.model, 'ckpt'):
                ckpt_dict = self.model.ckpt
                train_metrics = ckpt_dict.get("train_metrics", None)
                print("Train Metrics:", train_metrics)
            if hasattr(self.model, 'overrides') and 'imgsz' in self.model.overrides:
                self.input_size = self.model.overrides['imgsz']
            elif hasattr(self.model, 'train_args') and 'imgsz' in self.model.train_args:
                self.input_size = self.model.train_args['imgsz']
            else:
                self.input_size = 640  # Default value if not found

    def predict(self, image, device='cpu', task='detect'):
        if task not in ['detect', 'segment']:
            raise ValueError("Invalid task. Supported tasks are 'detect' and 'segment'.")

        if self.use_onnx:
            image_np = self.preprocess_image(image)
            input_name = self.model.get_inputs()[0].name
            outputs = self.model.run(None, {input_name: image_np})

            if task == 'detect':
                # Detection outputs are in outputs[0]
                detections = outputs[0][0]
                print("Detections:", detections.shape if detections is not None else 'None')
                return detections
            elif task == 'segment':
                detections, masks = self.model_onnx_seg_convert(outputs)
                print("Detections:", detections.shape if detections is not None else 'None',
                      "Masks:", masks.shape if masks is not None else 'None')
                return detections, masks
        else:
            results = self.model.predict(source=image, device=device)
            if task == 'detect':
                detections = results[0].boxes.data.cpu().numpy()
                print("Detections:", detections.shape if detections is not None else 'None')
                return detections
            elif task == 'segment':
                detections = results[0].boxes.data.cpu().numpy()
                masks = results[0].masks.data.cpu().numpy() if results[0].masks else None
                print("Detections:", detections.shape if detections is not None else 'None',
                      "Masks:", masks.shape if masks is not None else 'None')
                return detections, masks

    def preprocess_image(self, image):
        img_size = self.input_size
        if isinstance(image, Image.Image):
            image = np.array(image)

        # Original image dimensions
        h0, w0 = image.shape[:2]
        self.original_image_shape = (h0, w0)

        # Resize image with aspect ratio maintained
        r = min(img_size / h0, img_size / w0)
        self.scale_ratio = r

        h_new, w_new = int(h0 * r), int(w0 * r)
        image_resized = cv2.resize(image, (w_new, h_new), interpolation=cv2.INTER_LINEAR)

        # Compute padding to maintain input_size dimensions
        pad_h = img_size - h_new
        pad_w = img_size - w_new
        pad_top = pad_h // 2
        pad_bottom = pad_h - pad_top
        pad_left = pad_w // 2
        pad_right = pad_w - pad_left
        self.pad = (pad_left, pad_top)  # (pad_x, pad_y)

        # Add padding
        image_padded = cv2.copyMakeBorder(
            image_resized,
            pad_top,
            pad_bottom,
            pad_left,
            pad_right,
            cv2.BORDER_CONSTANT,
            value=(114, 114, 114)
        )

        # Convert to numpy array and normalize
        image_np = image_padded.astype(np.float32) / 255.0
        # Convert from HWC to CHW format
        image_np = np.transpose(image_np, (2, 0, 1))
        # Add batch dimension
        image_np = np.expand_dims(image_np, axis=0)
        return image_np

    def model_onnx_seg_convert(self, outputs):
        # Extract model outputs
        output0 = outputs[0]  # Shape: (batch_size, num_attributes, num_predictions)
        output1 = outputs[1]  # Shape: (batch_size, num_masks, mask_height, mask_width)

        if output0 is None or output1 is None:
            print("Model outputs are None.")
            return np.empty((0, 6)), []

        # Transpose output0 to get shape (num_predictions, num_attributes)
        output0 = output0[0].T  # Shape (num_predictions, num_attributes): (8400, 116)
        output1 = output1[0]    # Shape (num_masks, mask_height, mask_width): (32, 160, 160)

        print(f"Output0 Shape: {output0.shape}")
        print(f"Output1 Shape: {output1.shape}")

        # Calculate number of classes and masks
        num_predictions, num_attributes = output0.shape
        num_masks = output1.shape[0] # 32
        num_classes = num_attributes - 4 - num_masks  # num_classes = 80

        if num_classes <= 0:
            print("Invalid number of classes.")
            return np.empty((0, 6)), []

        print(f"Number of classes: {num_classes}, Number of masks: {num_masks}")

        # Extract boxes, class scores, and mask coefficients
        boxes = output0[:, :4]
        class_scores = output0[:, 4:4 + num_classes]  # Since num_classes = 80
        mask_coeffs = output0[:, 4 + num_classes:]    # Remaining attributes are mask coefficients

        print(f"Boxes shape: {boxes.shape}")
        print(f"Class scores shape: {class_scores.shape}")
        print(f"Mask coefficients shape: {mask_coeffs.shape}")

        # Apply sigmoid to class scores coefficients
        class_scores = self.sigmoid(class_scores)

        # Get class IDs and confidences
        class_ids = np.argmax(class_scores, axis=1)
        confidences = np.max(class_scores, axis=1)

        # Filter detections by confidence threshold
        conf_threshold = 0.5  # Adjust as needed
        keep = confidences > conf_threshold
        if np.sum(keep) == 0:
            print("No detections above confidence threshold.")
            return np.empty((0, 6)), []

        boxes = boxes[keep]
        confidences = confidences[keep]
        class_ids = class_ids[keep]
        mask_coeffs = mask_coeffs[keep]

        # Convert boxes from [x_center, y_center, width, height] to [x_min, y_min, x_max, y_max]
        boxes[:, 0] -= boxes[:, 2] / 2  # x_min
        boxes[:, 1] -= boxes[:, 3] / 2  # y_min
        boxes[:, 2] += boxes[:, 0]      # x_max
        boxes[:, 3] += boxes[:, 1]      # y_max

        # Adjust boxes based on scaling ratio and padding
        boxes[:, [0, 2]] -= self.pad[0]  # Subtract padding on x axis
        boxes[:, [1, 3]] -= self.pad[1]  # Subtract padding on y axis
        boxes[:, [0, 2]] /= self.scale_ratio
        boxes[:, [1, 3]] /= self.scale_ratio

        # Clip boxes to original image dimensions
        h0, w0 = self.original_image_shape
        boxes[:, [0, 2]] = boxes[:, [0, 2]].clip(0, w0)
        boxes[:, [1, 3]] = boxes[:, [1, 3]].clip(0, h0)

        print(f"Boxes after conversion and scaling:\n{boxes}")

        # Perform Non-Maximum Suppression (NMS)
        indices = self.nms(boxes, confidences, class_ids, iou_threshold=0.5)
        if len(indices) == 0:
            print("No detections after NMS.")
            return np.empty((0, 6)), []

        # Filter boxes, confidences, class_ids, and mask_coeffs based on NMS indices
        boxes = boxes[indices]
        confidences = confidences[indices]
        class_ids = class_ids[indices]
        mask_coeffs = mask_coeffs[indices]

        # Generate masks only for detections kept after NMS
        masks = self.generate_masks(mask_coeffs, output1)

        # Combine boxes, confidences, and class_ids into detections array
        detections = np.hstack((boxes, confidences[:, None], class_ids[:, None]))

        return detections, masks

    def nms(self, boxes, scores, class_ids, iou_threshold=0.5):
        # Perform class-aware NMS
        keep = []
        for cls in np.unique(class_ids):
            idxs = np.where(class_ids == cls)[0]
            cls_boxes = boxes[idxs]
            cls_scores = scores[idxs]

            x1 = cls_boxes[:, 0]
            y1 = cls_boxes[:, 1]
            x2 = cls_boxes[:, 2]
            y2 = cls_boxes[:, 3]

            areas = (x2 - x1) * (y2 - y1)
            order = cls_scores.argsort()[::-1]

            while order.size > 0:
                i = order[0]
                keep.append(idxs[i])
                xx1 = np.maximum(x1[i], x1[order[1:]])
                yy1 = np.maximum(y1[i], y1[order[1:]])
                xx2 = np.minimum(x2[i], x2[order[1:]])
                yy2 = np.minimum(y2[i], y2[order[1:]])

                w = np.maximum(0.0, xx2 - xx1)
                h = np.maximum(0.0, yy2 - yy1)
                inter = w * h
                union = areas[i] + areas[order[1:]] - inter
                iou = inter / (union + 1e-6)

                inds = np.where(iou <= iou_threshold)[0]
                order = order[inds + 1]

        return keep

    def generate_masks(self, mask_coeffs, mask_protos):
        num_detections = mask_coeffs.shape[0]
        num_masks, mask_height, mask_width = mask_protos.shape

        print(f"mask_coeffs shape: {mask_coeffs.shape}")
        print(f"mask_protos shape: {mask_protos.shape}")

        # Reshape mask prototypes to (num_masks, mask_height * mask_width)
        mask_protos_flat = mask_protos.reshape(num_masks, -1)  # Shape: (num_masks, mask_height * mask_width)

        # Generate masks: (num_detections, num_masks) @ (num_masks, mask_height * mask_width)
        masks = np.dot(mask_coeffs, mask_protos_flat)  # Shape: (num_detections, mask_height * mask_width)

        # Apply sigmoid activation
        masks = self.sigmoid(masks)

        # Reshape masks
        masks = masks.reshape(num_detections, mask_height, mask_width)

        # Upscale masks to original image size and apply threshold
        upscaled_masks = []
        for mask in masks:
            # Resize mask to original image size
            mask = cv2.resize(mask, (self.original_image_shape[1], self.original_image_shape[0]), interpolation=cv2.INTER_LINEAR)
            # Threshold the mask
            mask = (mask > 0.5).astype(np.uint8)
            upscaled_masks.append(mask)
        masks = np.array(upscaled_masks)

        return masks

    def sigmoid(self, x):
        # Numerically stable sigmoid
        return np.where(
            x >= 0,
            1 / (1 + np.exp(-x)),
            np.exp(x) / (1 + np.exp(x))
        )
```

The output using yolov8l-seg.onnx
![bicycle_detections_preview](https://github.com/user-attachments/assets/5f24681e-d92f-4de1-b058-3b9edfe74696)

The output using yolov8l-seg.pt
![bicycle_detections_preview](https://github.com/user-attachments/assets/b6270de4-a71a-4067-904f-7df0230c0468)

**Comment by glenn-jocher** [comment]
It seems like the issue might be related to differences in preprocessing or postprocessing between the ONNX and PyTorch models. Ensure that both models use the same preprocessing steps, including image resizing and normalization. Also, verify that the postprocessing logic, especially for masks and bounding boxes, is consistent. If the issue persists, consider checking the model's metadata and outputs for any discrepancies.

**Comment by curtis18** [comment]
@glenn-jocher Thank you so much. The ONNX results have improved. Is the problem caused by the downscaling and upscaling of ONNX? The accuracy of ONNX is not as high as that of the PT model.

![bicycle_detections_preview](https://github.com/user-attachments/assets/f5bbd53e-1856-46cb-b448-284ca2323a73)

**Comment by glenn-jocher** [comment]
The difference in accuracy could be due to precision loss during ONNX conversion or differences in how ONNX handles certain operations. Ensure consistent preprocessing and postprocessing steps for both models to minimize discrepancies.

**Comment by curtis18** [comment]
@glenn-jocher Thanks a lot.

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions, feel free to ask.
